### PR TITLE
Initial implemented RADIUS Protocol

### PR DESCRIPTION
**Pull request by takumin**
- Support Attributes Type String():
  - [RFC 2865 \- Remote Authentication Dial In User Service \(RADIUS\)](https://tools.ietf.org/html/rfc2865)
  - [RFC 2866 \- RADIUS Accounting](https://tools.ietf.org/html/rfc2866)
  - [RFC 2867 \- RADIUS Accounting Modifications for Tunnel Protocol Support](https://tools.ietf.org/html/rfc2867)
  - [RFC 2868 \- RADIUS Attributes for Tunnel Protocol Support](https://tools.ietf.org/html/rfc2868)
  - [RFC 2869 \- RADIUS Extensions](https://tools.ietf.org/html/rfc2869)
- Partial support EAP-Message attributes

### CODE REVIEW

**Comment by takumin** [comment]
@googlebot I signed it!

**Comment by takumin** [comment]
again...

@googlebot I signed it!

File: layers/radius.go [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+	if len(data) < radiusMinimumRecordSizeInBytes {
+		df.SetTruncated()
+		return fmt.Errorf("RADIUS length %d too short", len(data))
+	}
+
+	radius.BaseLayer = BaseLayer{Contents: data}
+
+	radius.Code = RADIUSCode(data[0])
+	radius.Identifier = RADIUSIdentifier(data[1])
+	radius.Length = RADIUSLength(binary.BigEndian.Uint16(data[2:4]))
```
Comment (tomfitzhenry) [inline layers/radius.go:389]: What should happen if radius.Length < radiusMinimumRecordSizeInBytes ?

File: layers/radius.go [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+	copy(radius.Authenticator[:], data[4:20])
+
+	if len(data) == radiusMinimumRecordSizeInBytes {
+		return nil
+	}
+
+	reader := bytes.NewReader(data[20:])
```
Comment (tomfitzhenry) [inline layers/radius.go:396]: "Octets outside the range of the Length field MUST be treated as padding and ignored on reception." https://tools.ietf.org/html/rfc2865#page-14

By using the entirety of `data`, does this code use octets outside of the range of the Length field?

File: layers/radius.go [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+
+	radius.Code = RADIUSCode(data[0])
+	radius.Identifier = RADIUSIdentifier(data[1])
+	radius.Length = RADIUSLength(binary.BigEndian.Uint16(data[2:4]))
+	copy(radius.Authenticator[:], data[4:20])
+
+	if len(data) == radiusMinimumRecordSizeInBytes {
+		return nil
+	}
+
```
Comment (tomfitzhenry) [inline layers/radius.go:395]: What should happen if radius.Length > len(data)?

"If the packet is shorter than the Length field indicates, it MUST be silently discarded." https://tools.ietf.org/html/rfc2865#page-14

File: layers/radius.go [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+	for {
+		header = header[:2]
+
+		if _, err := reader.Read(header); err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("RADIUS unknown error: %s", err)
+		}
+
+		value = value[:header[1]-2]
```
Comment (tomfitzhenry) [inline layers/radius.go:408]: Consider using variables for `header[1]` to make it clearer that this is a length field.

File: layers/radius.go (hunk 2 above)
Comment (tomfitzhenry) [reply layers/radius.go:396]: I think this should be checked and truncated, e.g.:
```
if len(data) > radius.Length {
  df.setTruncated()
  data = data[:radius.Length]
}
```

File: layers/radius.go (hunk 3 above)
Comment (tomfitzhenry) [reply layers/radius.go:395]: I think your code handles this fine: line 402's `err == io.EOF` will cause attribute parsing to early-exit.

File: layers/radius.go (hunk 1 above)
Comment (takumin) [reply layers/radius.go:389]: `RFC 2865 3. Packet Format` does not have a definition of less than 20 bytes, so change it to execute` df.SetTruncated () `.

https://tools.ietf.org/html/rfc2865#page-15

Comment (takumin) [reply layers/radius.go:389]: Implement the following code on line 370

```go
	if int(radius.Length) < radiusMinimumRecordSizeInBytes {
		df.SetTruncated()
		data = data[:radiusMinimumRecordSizeInBytes]
	}
```

File: layers/radius.go (hunk 2 above)
Comment (takumin) [reply layers/radius.go:396]: Thank you for your advice! I will implement it as you commented.

File: layers/radius.go (hunk 1 above)
Comment (takumin) [reply layers/radius.go:389]: I think the code below is appropriate, but how about it?

```go
	if int(radius.Length) < radiusMinimumRecordSizeInBytes {
		df.SetTruncated()
		return fmt.Errorf("RADIUS length %d too short", radius.Length)
	}
```

File: layers/radius.go (hunk 3 above)
Comment (takumin) [reply layers/radius.go:395]: When I added a test case based on your advice, there was a problem that was difficult to control, so I decided to rewrite the code drastically.

**Comment by takumin** [comment]
@tomfitzhenry 
The length check has been enhanced and test cases have been added.
Please review again!
Thank you very much

**Review by tomfitzhenry — APPROVED** [review]
LGTM!

**Comment by gconnell** [comment]
THanks a ton for the submission!

**Pull request merged**